### PR TITLE
Fix JsonItemExporter puts lone comma in the output if encoder fails

### LINKED ISSUE #3090: JsonItemExporter puts lone comma in the output if encoder fails
If `JsonItemExporter` is unable to encode the item, it still writes a delimiter (comma) to the output file. Here is a sample spider:

```
# -*- coding: utf-8 -*-
import datetime
import scrapy

class DummySpider(scrapy.Spider):
    name = 'dummy'
    start_urls = ['http://example.org/']

    def parse(self, response):
        yield {'date': datetime.date(2018, 1, 1)}
        yield {'date': datetime.date(1234, 1, 1)}
        yield {'date': datetime.date(2019, 1, 1)})
```

Encoding the second items fails:

```
2018-01-25 09:05:57 [scrapy.utils.signal] ERROR: Error caught on signal handler: <bound method ?.item_scraped of <scrapy.extensions.feedexport.FeedExporter object at 0x7fcbfbd81250>>
Traceback (most recent call last):
  File "/home/pasmen/SW/Python/data-sci-env/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/home/pasmen/SW/Python/data-sci-env/local/lib/python2.7/site-packages/pydispatch/robustapply.py", line 55, in robustApply
    return receiver(*arguments, **named)
  File "/home/pasmen/SW/Python/data-sci-env/local/lib/python2.7/site-packages/scrapy/extensions/feedexport.py", line 224, in item_scraped
    slot.exporter.export_item(item)
  File "/home/pasmen/SW/Python/data-sci-env/local/lib/python2.7/site-packages/scrapy/exporters.py", line 130, in export_item
    data = self.encoder.encode(itemdict)
  File "/usr/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
  File "/home/pasmen/SW/Python/data-sci-env/local/lib/python2.7/site-packages/scrapy/utils/serialize.py", line 22, in default
    return o.strftime(self.DATE_FORMAT)
ValueError: year=1234 is before 1900; the datetime strftime() methods require year >= 1900
```

The output looks like this:

```
[
{"date": "2018-01-01"},
,
{"date": "2019-01-01"}
]
```

This seems not to be a valid JSON file as e.g. `json.load()` and `jq` fail to parse it.

I think the problem is in [`export_item`](https://github.com/scrapy/scrapy/blob/108f8c4fd20a47bb94e010e9c6296f7ed9fdb2bd/scrapy/exporters.py#L123) method of `JsonItemExporter` class where it outputs the comma before decoding the item. The correct approach would be to try to decode the item (possibly with other needed operations) and perform the write atomically.

### PR DESCRIPTION
**Pull request by adnan-awan**
Continuation of #3111.

Resolves #3090, closes #3111.

### CODE REVIEW

File: scrapy/exporters.py [hunk 1]
```diff
@@ -142,14 +149,15 @@ def finish_exporting(self):
...
-            self._beautify_newline()
-        itemdict = dict(self._get_serialized_fields(item))
-        data = self.encoder.encode(itemdict)
-        self.file.write(to_bytes(data, self.encoding))
+        # catch exception and don't add empty item for export
+        try:
+            itemdict = dict(self._get_serialized_fields(item))
+            data = to_bytes(self.encoder.encode(itemdict), self.encoding)
+        except TypeError as e:
+            raise e
```
Comment (wRAR) [inline scrapy/exporters.py:157]: I don't think this catch+reraise does anything?

File: tests/test_exporters.py [hunk 2]
```diff
@@ -599,6 +599,21 @@ def test_two_items(self):
...
+        i2 = TestItem(
+            name="Maria", age=1j
+        )  # Invalid datetimes didn't consistently fail between Python versions
+        i3 = TestItem(name="Jesus", age="44")
+        self.ie.start_exporting()
+        self.ie.export_item(i1)
+        self.assertRaises(TypeError, self.ie.export_item, i2)
+        self.ie.export_item(i3)
+        self.ie.finish_exporting()
+        print(f"values are=={self.output.getvalue()}")
```
Comment (wRAR) [inline tests/test_exporters.py:613]: I guess this is a debug print?

File: scrapy/exporters.py (hunk 1 above)
Comment (adnan-awan) [reply scrapy/exporters.py:157]: @wRAR  thanks for your comment. 

Actually, [this](https://github.com/scrapy/scrapy/pull/5952/commits/0bd481de7cec4540aad8b9cfae896519fd9b2016#diff-b4b7b51f903343e1ac43a4655e42263010d0631f9d1702eb5058cbd5d5402d47R608) test expect the TypeError. Do you've something better in your mind for its replacement?.

Comment (wRAR) [reply scrapy/exporters.py:157]: But as this code does nothing then removing it won't break the test.

**Review by wRAR — APPROVED** [review]

**Comment by kmike** [comment]
Thanks @adnan-awan!

**Pull request merged**